convert-mpt-hf-to-gguf.py

```diff
@@ -0,0 +1,264 @@
+#!/usr/bin/env python3
+# HF mpt--> gguf conversion
+
+from __future__ import annotations
+
+import argparse
+import json
+import os
+import struct
+import sys
+from pathlib import Path
+from typing import Any
+
+import numpy as np
+import torch
+from transformers import AutoTokenizer  # type: ignore[import]
+
+if 'NO_LOCAL_GGUF' not in os.environ:
+    sys.path.insert(1, str(Path(__file__).parent / 'gguf-py' / 'gguf'))
+import gguf
+
+# ref: https://github.com/openai/gpt-2/blob/master/src/encoder.py
+
+
+def bytes_to_unicode():
+    """
+    Returns list of utf-8 byte and a corresponding list of unicode strings.
+    The reversible bpe codes work on unicode strings.
+    This means you need a large # of unicode characters in your vocab if you want to avoid UNKs.
+    When you're at something like a 10B token dataset you end up needing around 5K for decent coverage.
+    This is a significant percentage of your normal, say, 32K bpe vocab.
+    To avoid that, we want lookup tables between utf-8 bytes and unicode strings.
+    And avoids mapping to whitespace/control characters the bpe code barfs on.
+    """
+    bs = list(range(ord("!"), ord("~")+1))+list(range(ord("¡"), ord("¬")+1))+list(range(ord("®"), ord("ÿ")+1))
+    cs = bs[:]
+    n = 0
+    for b in range(2**8):
+        if b not in bs:
+            bs.append(b)
+            cs.append(2**8+n)
+            n += 1
+    return dict(zip(bs, (chr(n) for n in cs)))
```

Comment:
Can we use `transformers.models.gpt2.tokenization_gpt2.bytes_to_unicode` instead of copy-pasting it here? (see https://github.com/ggerganov/llama.cpp/pull/3405/commits/846c51feacfc87fbcb2a7f9bc411de714e09256e)

Reply:
Yes, I could clean it up before merge, but I was under impression that this code would be wiped out anyway later by replacing it with the more concise version implemented by @goerch in PR #3252 (which is already merged into master). I was hoping that this final cleanup of the convert script could be done after merging this PR, rather than before, because doing otherwise would require me to sync up my mpt branch with the 35 commits from PR #3252 which would then redundantly be included in this PR as well (LMK if this is a more desirable approach anyway).

Reply:
I guess we could also close this PR, and I could prepare a new PR branch against the latest llama.cpp master, but apart from more effort, it is perhaps also not ideal for documenting history. My proposal would be to rather pull this PR into master, and then you could update/"bugfix" the convert script to match gpt-neox while applying the few required MPT-related diffs on top of it.

Reply:
Okay, I suppose I'll save my review of this script for the later PR.